File: functions/src/model/reservation_model.ts
export class ReservationModel {
    id:string = "";
    userId:string = "";
    restaurantId: string = "";
    checkIn: Date = new Date;
    occupationQty: number = 0;
    status: ReservationStatus = ReservationStatus.ATIVO;
}

export enum ReservationStatus {
    RESERVADO,
    AGUARDANDO,
    ATIVO,
    FINALIZADO,
    CANCELADO,
  }

export function convertReservationFromDB(reservationFromDB: any) {
    var reservation = new ReservationModel;
  
    if (reservationFromDB) {
        reservation.id = reservationFromDB.id;
        reservation.userId = reservationFromDB.data()["userId"];
        reservation.restaurantId = reservationFromDB.data()["restaurantId"];
        reservation.checkIn = reservationFromDB.data()["checkin"];
        reservation.occupationQty = reservationFromDB.data()["occupationQty"];
        reservation.status = getStatusEnum(reservationFromDB.data()["status"]);
    }
    
    return reservation;
  }

  function getStatusEnum(statusFromDB: string) {
    switch (statusFromDB) {
        case "RESERVADO":
            return ReservationStatus.RESERVADO;
        case "AGUARDANDO":
            return ReservationStatus.AGUARDANDO;
        case "ATIVO":
            return ReservationStatus.ATIVO;
        case "FINALIZADO":
            return ReservationStatus.FINALIZADO;
        case "CANCELADO":
            return ReservationStatus.CANCELADO;
        default:
            return ReservationStatus.ATIVO;
    }
  }

File: lib/app/data/model/reservation_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:sit_eat/app/data/model/enum/reservation_status_enum.dart';

class ReservationModel {
  String id;
  String userId;
  String restaurantId;
  Timestamp checkIn;
  int occupationQty;
  ReservationStatus status;

  ReservationModel({
    this.id,
    this.userId,
    this.restaurantId,
    this.checkIn,
    this.occupationQty,
    this.status,
  });

  ReservationModel.fromSnapshot(DocumentSnapshot reservation)
      : id = reservation.id,
        userId = reservation.data()["userId"],
        restaurantId = reservation.data()["restaurantId"],
        checkIn = reservation.data()["checkin"],
        occupationQty = reservation.data()["occupationQty"],
        status = ReservationStatus.values.where((status) => status.toUpper == reservation.data()["status"]).first;
}

List<ReservationModel> convertReservationsFromDB(QuerySnapshot reservationsFromDB) {
  List<ReservationModel> reservations = <ReservationModel>[];
  reservationsFromDB.docs.forEach((restaurant) {
    reservations.add(ReservationModel.fromSnapshot(restaurant));
  });
  return reservations;
}
